### PR TITLE
Unregister replication instance

### PR DESCRIPTION
**Pull request by as51340**
PR enables unregistering replication instances. Coordinator instances cannot be yet removed. This will be added once coordinators are part of all instances. If instance was replica, main will stop replicating to this replica while the state of the rest of the cluster will be preserved. If instance was main, we will initiate failover and try to choose new main.

[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [ ] Provide the full content or a guide for the final git message

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here, including added/changed clauses
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

**Review by andrejtonev — COMMENTED** [review]
Looks good!

File: src/coordination/coordinator_instance.cpp [hunk 1]
```diff
@@ -308,6 +308,39 @@ auto CoordinatorInstance::RegisterReplicationInstance(CoordinatorClientConfig co
   return RegisterInstanceCoordinatorStatus::SUCCESS;
 }
 
+auto CoordinatorInstance::UnregisterReplicationInstance(std::string instance_name)
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:311]: Not yet, we have support for log store but it is not synced with state machine. We want to solve first replication things and then go back to raft

File: src/coordination/coordinator_client.cpp [hunk 2]
```diff
@@ -121,5 +126,19 @@ auto CoordinatorClient::SendSwapMainUUIDRpc(const utils::UUID &uuid) const -> bo
...
+auto CoordinatorClient::SendUnregisterReplicaRpc(std::string const &instance_name) const -> bool {
+  try {
+    auto stream{rpc_client_.Stream<UnregisterReplicaRpc>(instance_name)};
+    if (!stream.AwaitResponse().success) {
+      spdlog::error("Failed to receive successful RPC response for unregistering replica!");
+      return false;
+    }
+    return true;
+  } catch (rpc::RpcFailedException const &) {
+    spdlog::error("Failed to unregister replica!");
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_client.cpp:138]: What do you mean by ignored? We return false from coordinator_client which is then checked later.
> The only way to detect that an instance is down is through the frequent check?

yes

File: src/coordination/coordinator_instance.cpp [hunk 3]
```diff
@@ -308,6 +308,39 @@ auto CoordinatorInstance::RegisterReplicationInstance(CoordinatorClientConfig co
...
+  if (is_replica) {
+    auto curr_main = std::ranges::find_if(repl_instances_, &ReplicationInstance::IsMain);
+    MG_ASSERT(curr_main != repl_instances_.end(), "There must be a main instance when unregistering a replica");
+    if (!curr_main->SendUnregisterReplicaRpc(instance_name)) {
+      return UnregisterInstanceCoordinatorStatus::RPC_FAILED;
+    }
+    std::erase_if(repl_instances_, name_matches);
+  } else {
+    std::erase_if(repl_instances_, name_matches);
+    TryFailover();
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:338]: lock is taken, I think that should be fine

[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:338]: but also when we delete it, thread is destroyed. For replica callbacks I think nothing bad will happen and for main we are thinking about forbidding unregistering current main instance because if you take a look at else branch there is TryFailover call. We cannot guarantee that failover will succeed hence after unregistering we could get in a cluster state without main instance => I don't like that, unregistering instance is too rare and too simple operation to produce such result. If you want to unregister main instance, kill it and then unregister. What do you say?

File: src/coordination/coordinator_instance.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:311]: consensus not needed, we will request leadership and if leader manages to append to log, raft provides other guarantees.

File: src/coordination/coordinator_instance.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrejtonev) [reply src/coordination/coordinator_instance.cpp:338]: I'm thinking about the stream itself. 
We could be sending the RPC message, serializing/deserializing it, calling handlers and so on. Not sure that it safe to do. That part is not under lock.
So what happens if we destroy an instance while also communicating with it in another thread?

File: src/coordination/coordinator_client.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrejtonev) [reply src/coordination/coordinator_client.cpp:138]: Ignored in the sense that the ping failed function is never called. And the only way to update the alive status is through frequent check.

File: src/coordination/coordinator_instance.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (antoniofilipovic) [reply src/coordination/coordinator_instance.cpp:338]: I agree. We could have a stream running, and then we erase instance from our list, and since there is callback left to do, we will continue in the function finding instance and that will kill coordinator. I think that can happen, we can get on call if needed about this, it is real edge case as I see it

File: tests/e2e/high_availability_experimental/coord_cluster_registration.py [hunk 4]
```diff
@@ -280,5 +284,129 @@ def check_coordinator2():
     mg_sleep_and_assert(expected_cluster_not_shared, check_coordinator2)
 
 
+# TODO: (andi) Test when dealing with distributed coordinators that you can register on one coordinator and unregister from any other coordinator
+@pytest.mark.parametrize(
+    "kill_instance",
+    [True, False],
+)
+def test_unregister_replicas(kill_instance):
+    safe_execute(shutil.rmtree, TEMP_DIR)
```
Comment (antoniofilipovic) [inline tests/e2e/high_availability_experimental/coord_cluster_registration.py:293]: Can you add steps here what we do in each step, it is easier to get intention just from reading description

File: src/coordination/coordinator_client.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_client.cpp:138]: Why would I call ping failed if unregister replica rpc message is sent?

**Review by antoniofilipovic — COMMENTED** [review]
One comment, that is it from me

File: src/coordination/coordinator_instance.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:338]: All code should be written in a way that if the communication to instance fails, we can continue operating. I think it should be ok

[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:338]: but good point, I will check further if this can happen

**Review by antoniofilipovic — APPROVED** [review]

**Comment by as51340** [comment]
@kgolubic Release note: Users can unregister instances from the cluster. Alive main cannot be unregistered due to possible cluster state issues that can happen.
